### ISSUE TITLE
Case insensitive check for Unicode

### ISSUE THREAD

**Issue by Alexey-T** [andgineer/TRegExpr#223]
I git comment

-It is also not fully correct. E.g. K (U+212a) should match K or k case insensitively

So our approach has weak place. We match case insensitive using 2 arrays- to lowercase, to uppercase.
How to fix?

**Comment by AmigoJack** [comment]
Covering this for **every** letter is possible, but results in huge arrays. https://www.unicode.org/Public/UCD/latest/ucd/UnicodeData.txt covers all Unicode graphemes/letters. As described in https://www.unicode.org/L2/L1999/UnicodeData.html the last 3 columns will tell you if the letter has an uppercase/lowercase/titlecase counterpart. For example:

1. > `0030;DIGIT ZERO;Nd;0;EN;;0;0;0;N;;;;;`
   - Last 3 columns are empty, so the codepoint U+0030 is uninteresting for case insensitivity.

2. > `0044;LATIN CAPITAL LETTER D;Lu;0;L;;;;;N;;;;0064;`
   - The "lowercase mapping" column is filled: codepoint U+0044's lowercase counterpart is U+0064.

3.  > `212A;KELVIN SIGN;Lu;0;L;004B;;;;N;DEGREES KELVIN;;;006B;`
    - The "lowercase mapping" column is filled: codepoint U+212A's lowercase counterpart is U+006B.

4. > `01C5;LATIN CAPITAL LETTER D WITH SMALL LETTER Z WITH CARON;Lt;0;L;<compat> 0044 017E;;;;N;LATIN LETTER CAPITAL D SMALL Z HACEK;;01C4;01C6;01C5`
   - Uppercase mapping is U+01C4.
   - Lowercase mapping is U+01C6.
   - Titlecase mapping (when the letter is the first of a word) is U+01C5 (itself).

From those definitions I made arrays and lookup functions myself, covering all 2360 (BMP) + 520 (SMP) cases for turning one letter to the other. Plus 47 letters and how they should be normalized when trying to "fold" them for case insensitivity, too. It's written for Delphi 7, so searching for SMP characters would only work via `FindChar()` and not via `UniToLower()`. I included a full working program testing a few examples for lowercasing individual codepoints and showing how to use **uUnicodeCase.pas**, which would be something TRegExpr would need to include: [UnicodeCasing.zip](https://github.com/andgineer/TRegExpr/files/12674380/UnicodeCasing.zip)

**Comment by AmigoJack** [comment]
Yes. Capital and small letters are a concept not used in many languages/alphabets, and signs which look like capital and small variants are still different things and can't be exchanged (f.e. Katakanas). Have a read on https://en.wikipedia.org/wiki/Letter_case

**Comment by Alexey-T** [comment]
I didnt see your code yet. Im not at PC. If those char codes (which are changed by case operation) are located in few compact ranges, we can write simple change-case functions (3 of them). Can you write these functions?

**Comment by Alexey-T** [comment]
I see the code: very big work is done. good. now i ask: now much the speed loss is? today, TRegExpr lowercase/upcase are very fast (but they need 2 130Kb arrays: to lower, to upper).
will it be much slower if we replace funcs to your funcs?

if not much slower: where can we find 3 functions: to-lower, to-upper, to-title? I mean funcs which change WideChar to WideChar.

**Comment by Alexey-T** [comment]
this place is to change in regexpr.pas:
```
{$IFDEF FastUnicodeData}
function _UpperCase(Ch: REChar): REChar; {$IFDEF InlineFuncs}inline;{$ENDIF}
begin
  Result := CharUpperArray[Ord(Ch)];
end;

function _LowerCase(Ch: REChar): REChar; {$IFDEF InlineFuncs}inline;{$ENDIF}
begin
  Result := CharLowerArray[Ord(Ch)];
end;
```

**Comment by AmigoJack** [comment]
No, you haven't understood it throughly. Just because it's a function you don't need to care for the result - its parameter is a reference, not only a value. You could do:
```
function _LowerCase( Ch: REChar ): REChar;
begin
  result:= Ch;
  uUnicodeCase.UniToLower( result );  // parameter is VAR for a reason
end;
```
The function result just tells you if there was a match. Both other functions act the same way. Be aware that in those functions I casted the `WideChar` parameter to `Word` (both being 16 bit) due to only acting on Delphi 7. But Unicode wise this only works for codepoints U+0000 to U+FFFF, ignoring anything beyond (like Emojis). I'm not sure how `Char` works on other Delphi or FPC versions - logic wise it must also be capable to store at least 20 bit (likely a `Cardinal` = 32 bit). In Delphi 7 `WideString` encodes text in UTF-16, so any codepoints beyond U+FFFF are stored in **two** `WideChar`s (which poses a logical problem to properly convert them to upper/lower/title case. Luckily Unicode's Supplementary Multilingual Plane (codepoints U+10000 to U+1FFFF) has only lower/upper case letters which should be encountered **very** rarily (I guess that's the reason why all SMP isn't even available/ignored in your `regexpr_unicodedata.pas`).

Speed wise it must be slower than your two giant arrays (at worst 177 jumps and 15 loop iterations for 1 character, at best 1 jump and 1 loop iteration), but on the contrary my approach should need much less memory/storage. If you want I can benchmark both (or at least write code for it). Sometimes I'm eager for speed, sometimes I want to avoid bloat - both approaches have their advantages.

Either way - codepoints beyond U+FFFF should be dealth with, also for the Unicode category (as "word character"). If picking one character out of a `String` in recent Delphi versions can go beyond U+FFFF then let me at least double two of your arrays, so they also go until U+1FFFF, and increase the category array up to U+E01EF (14x as big). We could provide compiler switches for people who really want all Unicode codepoints covered and don't care that they then have 256 KiB + 256 KiB + 896,5 KiB = 1,38 MiB of constant data.

**Comment by Alexey-T** [comment]
I got it, so I can write the patch.
one thing why I don't entisuast here. i don't understand the code, ie those 2 arrays, how you create them. w/o understanding it, i won't comply with idea.
